### PR TITLE
[CRT] Fixed _tstat64 which wrongly set executable bits on files. CORE…

### PR DESCRIPTION
**Pull request by Extravert-ir**
## Purpose

initdb.exe now able to start postgresql.exe

JIRA issue: [CORE-10904](https://jira.reactos.org/browse/CORE-10904)

## Proposed changes
This PR:
- fixes bug in _tstat64 where it wrongly set execution bits in st_mode
- replaces some functions with their tchar.h definitions in order to better handle UNICODE mode

### CODE REVIEW

**Comment by Extravert-ir** [comment]
I wanted to alter crt test to check this behaviour, but is seems that I have to post it to wine

**Comment by HBelusca** [comment]
You can always create a new (or alter) an "apitest" (instead of a "winetest") if you don't want to send the tests to wine.

File: sdk/lib/crt/stdio/stat64.c [hunk 1]
```diff
@@ -14,29 +14,28 @@ HANDLE fdtoh(int fd);
...
+#define TCSIZE_BITS sizeof(_TCHAR)*8
 
-#define EXE ((TOULL('e')<<(2*TCSIZE)) | (TOULL('x')<<TCSIZE) | (TOULL('e')))
-#define BAT ((TOULL('b')<<(2*TCSIZE)) | (TOULL('a')<<TCSIZE) | (TOULL('t')))
-#define CMD ((TOULL('c')<<(2*TCSIZE)) | (TOULL('m')<<TCSIZE) | (TOULL('d')))
-#define COM ((TOULL('c')<<(2*TCSIZE)) | (TOULL('o')<<TCSIZE) | (TOULL('m')))
+#define EXE (((TCHAR4)('e')<<(2*TCSIZE_BITS)) | ((TCHAR4)('x')<<TCSIZE_BITS) | ((TCHAR4)('e')))
+#define BAT (((TCHAR4)('b')<<(2*TCSIZE_BITS)) | ((TCHAR4)('a')<<TCSIZE_BITS) | ((TCHAR4)('t')))
+#define CMD (((TCHAR4)('c')<<(2*TCSIZE_BITS)) | ((TCHAR4)('m')<<TCSIZE_BITS) | ((TCHAR4)('d')))
+#define COM (((TCHAR4)('c')<<(2*TCSIZE_BITS)) | ((TCHAR4)('o')<<TCSIZE_BITS) | ((TCHAR4)('m')))
```
Comment (HBelusca) [inline sdk/lib/crt/stdio/stat64.c]: What is that horror?

File: sdk/lib/crt/stdio/stat64.c [hunk 2]
```diff
@@ -68,10 +67,13 @@ int CDECL _tstat64(const _TCHAR *path, struct __stat64 *buf)
...
+    if (plen > 6 && path[plen-4] == __T('.'))  /* shortest exe: "\x.exe" */
     {
-      unsigned int ext = tolower(path[plen-1]) | (tolower(path[plen-2]) << 8) |
-                                 (tolower(path[plen-3]) << 16);
+
+      TCHAR4 ext = (TCHAR4)_totlower(path[plen-1])
+                   | ((TCHAR4)_totlower(path[plen-2]) << TCSIZE_BITS)
+                   | ((TCHAR4)_totlower(path[plen-3]) << 2*TCSIZE_BITS);
+
       if (ext == EXE || ext == BAT || ext == CMD || ext == COM)
```
Comment (HBelusca) [inline sdk/lib/crt/stdio/stat64.c]: Using a _tcsnicmp call with the strings L".exe", etc... L".com" would possibly be more readable.

Comment (ThFabba) [reply sdk/lib/crt/stdio/stat64.c]: This is based on Wine code, which does exactly this shift stuff

File: sdk/lib/crt/stdio/stat64.c [hunk 3]
```diff
@@ -14,29 +14,28 @@ HANDLE fdtoh(int fd);
 #define TCHAR4 ULONG
 #endif
 
-#define TCSIZE sizeof(_TCHAR)
-#define TOULL(x) ((TCHAR4)(x))
+#define TCSIZE_BITS sizeof(_TCHAR)*8
```
Comment (ThFabba) [inline sdk/lib/crt/stdio/stat64.c]: Oh, missing 8! Nice catch!

Comment (ThFabba) [inline sdk/lib/crt/stdio/stat64.c]: It's good practice to put parentheses around the macro expressions so you don't get surprised by operator precedence issues when using them

File: sdk/lib/crt/stdio/stat64.c (hunk 2 above)
Comment (Extravert-ir) [reply sdk/lib/crt/stdio/stat64.c]: Is there a reason why it was done this way?

File: sdk/lib/crt/stdio/stat64.c (hunk 3 above)
Comment (Extravert-ir) [reply sdk/lib/crt/stdio/stat64.c]: Fixed

File: sdk/lib/crt/stdio/stat64.c (hunk 2 above)
Comment (ThFabba) [reply sdk/lib/crt/stdio/stat64.c]: Well, this code appears to be at least 16 years old. So even if there was a reason, it's probably not valid anymore.
That said, it's usually best for us to keep the code in sync with Wine unless we have a specific reason not to. That means less maintenance effort for us.

**Pull request merged**